### ISSUE TITLE
fork-thread may block when call ffi recv or send

### ISSUE THREAD

**Issue by evilbinary** [cisco/ChezScheme#238]
I made of socket fork thread recv and send data, But i found that it may block in recv or send.Here is my code.

```` 
(import (net socket-ffi)
	(cffi cffi) )

(define socket-fd 0)
(define n 0)
(define buff (cffi-alloc 4096))

(cffi-log #t)

(set! socket-fd (socket AF_INET SOCK_STREAM 0))
(display  (format "socket-fd=~a\n" socket-fd))

(define server-addr (make-sockaddr-in AF_INET INADDR_ANY 8000))

(define bind-ret (bind socket-fd  server-addr 16))
(display  (format "bind-ret=~a\n" bind-ret))

(define listen-ret (listen socket-fd 10))

(display  (format "listen-ret=~a\n" listen-ret))

(define (make-thread fd)
  (fork-thread
   (lambda () 
     (let loop2 ((buf (cffi-alloc 4096)))
       (printf "recv ret=~a\n" (recv fd buf 1024 0))
       (printf "send ret=~a\n" (send fd "test你好\n" 6 0))
       ;;(printf "get ~a ~a\n" fd (cffi-string buf))
       ;;(sleep (make-time 'time-duration 100 0))
      (loop2 buf))
     )))

(let loop ()
  (printf "loop-thread ~a \n"  (get-thread-id))
  (let ((connect-fd (accept socket-fd 0 0)))
    (display (format "accept connect-fd=~a\n" connect-fd))    
    (make-thread connect-fd)
    (printf "fork-thread end\n")
    )
  (loop))

(close connect-fd)
(close socket-fd)

**Comment by akeep** [comment]
I think the problem here is unrelated to Chez Scheme.

From the ``recv`` documentation:

> If no messages are available at the socket, the receive call waits for a message to arrive, unless the socket is nonblocking (see fcntl(2)) in which case the value -1 is returned and the external variable errno set to EAGAIN.  The receive calls normally return any data available, up to the requested amount, rather than waiting for receipt of the full amount requested; this behavior is affected by the socket-level options SO_RCVLOWAT and SO_RCVTIMEO described in getsockopt(2).

I tested this out by using ``fcntl`` to get the flags currently set on the accepted socket (command ``F_GETFL``), and found that the socket descriptor returned is not marked as non-blocking, which is why ``recv`` and ``send`` and blocking.  You should be able to get the behavior you want by using ``fcntl`` with ``F_SETFL`` on the socket descriptor returned by ``accept``.  (Note that if you set it on the original socket descriptor from ``socket``, the accept will not block, it will simply return an error message if no connections have been requested.)

**Comment by evilbinary** [comment]
it's have two socket connects.I just write a server with c.It dosen't block.I think the fork-thread may block in  recv or send function with mutithread.
Here is the code:

```` #include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<errno.h>
#include<sys/types.h>
#include<sys/socket.h>
#include<netinet/in.h>
#include <pthread.h>

#define MAXLINE 4096

pthread_t main_tid;  

void print_ids(const char *str)  
{  
    pid_t pid;
    pthread_t tid;  
    pid = getpid();  
    tid = pthread_self();  
    printf("%s pid: %u tid: %u (0x%x)/n",  
                str,  
                (unsigned int)pid,  
                (unsigned int)tid,  
                (unsigned int)tid);  
}  

void *func(void *arg)  
{
	char buff[4096];
	int n;	 
	int connfd=(int)arg; 
	print_ids("new  thread:");  
	for(;;){
		n = recv(connfd, buff, MAXLINE, 0);
		buff[n] = '\0';
		printf("recv msg from client: %s\n", buff);
		send(connfd,buff,6,0);
	}
    return ((void *)0);  
} 

int 
main(int argc, char **argv)
{
	int		listenfd  , connfd;
	struct sockaddr_in servaddr;
	char		buff      [4096];
	int		n;

	if ((listenfd = socket(AF_INET, SOCK_STREAM, 0)) == -1) {
		printf("create socket error: %s(errno: %d)\n", strerror(errno), errno);
		exit(0);
	}
	memset(&servaddr, 0, sizeof(servaddr));
	servaddr.sin_family = AF_INET;
	servaddr.sin_addr.s_addr = htonl(INADDR_ANY);
	servaddr.sin_port = htons(8000);

	if (bind(listenfd, (struct sockaddr *)&servaddr, sizeof(servaddr)) == -1) {
		printf("bind socket error: %s(errno: %d)\n", strerror(errno), errno);
		exit(0);
	}
	if (listen(listenfd, 10) == -1) {
		printf("listen socket error: %s(errno: %d)\n", strerror(errno), errno);
		exit(0);
	}
	printf("======waiting for client's request======\n");
	while (1) {
		if ((connfd = accept(listenfd, (struct sockaddr *)NULL, NULL)) == -1) {
			printf("accept socket error: %s(errno: %d)", strerror(errno), errno);
			continue;
		}
		int err;  
    	err = pthread_create(&main_tid, NULL, func, connfd);
    	if(err != 0){  
        	printf("create thread error: %s/n",strerror(err));  
        	return 1;  
    	}  
		//close(connfd);
	}

	close(listenfd);

	return EXIT_SUCCESS;
}

**Comment by evilbinary** [comment]
I finally found that Sactivate_thread  and   Sdeactivate_thread should be enclosed ffi function when call.

**Comment by Saigut** [comment]
@evilbinary Hi, how did you fix your scheme code, can you show it?

**Comment by evilbinary** [comment]
@Saigut  是这样的，你把ffi的c函数，比如void test_read(){Sactivate_thread Sdeactivate_thread

**Comment by Saigut** [comment]
谢谢！

**Comment by evilbinary** [comment]
@Saigut  you solved you problem right?

**Comment by Saigut** [comment]
Solved #127

**Comment by evilbinary** [comment]
This issue ref #267 #213 #64 #58
